### PR TITLE
Refactor parachains tracker

### PR DESCRIPTION
**Pull request by AndreiEres**
We added a lot of new features to the parachain-tracer last time. This made it difficult to understand. We also had a lack of testing in the module. 

### What's been done in this refactoring
- The tracker lost some tied fields, like storage and RPC wrappers.
- Storage and RPC wrappers have been moved as dependencies in method calls, making them easier to mock in testing.
- Improved test coverage. 

### What may be done in the future
- Async backing should be properly supported. See https://github.com/paritytech/polkadot-introspector/issues/537
- Storage and RPC wrappers may be merged, adding some cache layer because we do too many same RPC queries.

### CODE REVIEW

File: parachain-tracer/src/utils.rs [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+	types::{AccountId32, Timestamp, H256},
+};
+use std::time::Duration;
+
+/// Returns a time difference between optional timestamps
+pub(crate) fn time_diff(lhs: Option<u64>, rhs: Option<u64>) -> Option<Duration> {
+	match (lhs, rhs) {
+		(Some(lhs), Some(rhs)) => Some(Duration::from_millis(lhs).saturating_sub(Duration::from_millis(rhs))),
+		_ => None,
+	}
```
Comment (vstakhov) [inline parachain-tracer/src/utils.rs:29]:
```suggestion
  Some(Duration::from_millis(lhs?).saturating_sub(Duration::from_millis(rhs?)))
```

**Review by vstakhov — APPROVED** [review]
It looks very good, especially with all these new tests.

File: parachain-tracer/src/main.rs [hunk 2]
```diff
@@ -215,21 +221,22 @@ impl ParachainTracer {
...
-								.await
+								let parent_number = new_head.relay_parent_number;
+								if let Err(e) =
+									tracker.inject_block(*relay_fork, parent_number, &mut rpc, &storage).await
 								{
 									error!("error occurred when processing block {}: {:?}", relay_fork, e);
 									std::process::exit(1);
-								};
+								}
+								if let Some(progress) = tracker.progress(&mut stats, &metrics, &storage).await {
```
[the remark this answers is not in the document]
Comment (AndreiEres) [reply parachain-tracer/src/main.rs:231]: Maybe there is a bad wording? In `progress` we process updates received with current block as well as change current `stats` and `metrics`.

File: parachain-tracer/src/tracker.rs [hunk 3]
```diff
@@ -15,901 +15,1156 @@
...
+		rpc: &mut impl TrackerRpc,
+	) -> color_eyre::Result<()> {
+		if self.current_candidate.is_backed() {
+			// We only process availability if our parachain is assigned to an availability core.
+			if let Some(core) = self.current_candidate.assigned_core {
+				self.current_candidate.current_availability_bits = extract_availability_bits_count(bitfields, core);
+				self.current_candidate.max_availability_bits =
+					self.validators_indices(block_hash, rpc).await?.len() as u32;
+
+				if self.current_candidate.is_data_available() {
```
Comment (sandreim) [inline parachain-tracer/src/tracker.rs:315]: These are the availability bits for a previous candidate. We need to account for async backing.

**Review by sandreim — COMMENTED** [review]
Nice work @AndreiEres . I appreciate the effort here. I think there are two things to do before merge:
- add a proper PR description which describes the goal of the refactoring and how it's achieved
- make sure async backing is supported

File: essentials/src/chain_events.rs [hunk 4]
```diff
@@ -85,9 +85,10 @@ pub struct SubxtDispute {
 }
 
 /// Dispute result as seen by subxt event
-#[derive(Debug, Clone, Copy, Serialize, Decode, Encode, PartialEq, Eq)]
+#[derive(Debug, Clone, Copy, Serialize, Decode, Encode, PartialEq, Eq, Default)]
 pub enum SubxtDisputeResult {
 	/// Dispute outcome is valid
+	#[default]
```
[the remark this answers is not in the document]
Comment (AndreiEres) [reply essentials/src/chain_events.rs:91]: Do you mean we should add it as a default?

File: parachain-tracer/src/tracker.rs [hunk 5]
```diff
@@ -15,901 +15,1156 @@
...
-			update.bitfield_health.available_count = self.current_candidate.current_av_bits;
-			update.bitfield_health.bitfield_count = self.current_candidate.bitfield_count;
+	async fn notify_candidate_state(
+		&self,
+		progress: &mut ParachainProgressUpdate,
+		stats: &mut impl Stats,
+		metrics: &impl PrometheusMetrics,
+		storage: &TrackerStorage,
+	) {
+		if self.current_candidate.is_idle() {
```
[the remark this answers is not in the document]
Comment (AndreiEres) [reply parachain-tracer/src/tracker.rs:428]: I think so, we change current candidate on `inject_block` stage. Here we just process metrics according to its current stage.

**Pull request merged**